File: solutions/[106]从中序与后序遍历序列构造二叉树.cpp
#include <vector>
#include <map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
		map<int, int> m;
		for (int i = 0; i < inorder.size(); ++i)
			m[inorder[i]] = i;
		int post_idx = postorder.size()-1;
		return dp(m, postorder, inorder, post_idx, 0, inorder.size());
	}
	TreeNode* dp(map<int, int>& m, vector<int>& postorder, vector<int>& inorder, int& post_idx, int in_left, int in_right) {
		if (in_left == in_right)
			return nullptr;

		TreeNode* root = new TreeNode(postorder[post_idx]);
		int idx = m[postorder[post_idx]];
		post_idx--;
		root->right = dp(m, postorder, inorder, post_idx, idx + 1, in_right);
		root->left = dp(m, postorder, inorder, post_idx, in_left, idx);

		return root;
	}
};

File: solutions/[106]从中序与后序遍历序列构造二叉树.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        val2p_in = dict()
        for i, val in enumerate(inorder):
            val2p_in[val] = i
        def build_tree(p_post, l_in, r_in):
            if l_in > r_in:
                return None
            root_val = postorder[p_post]

            p_root_in = val2p_in[root_val]
            root = TreeNode(root_val,
                            build_tree(p_post-(r_in-p_root_in)-1, l_in, p_root_in-1),
                            build_tree(p_post-1, p_root_in+1, r_in))
            return root
        
        return build_tree(len(postorder)-1, 0, len(inorder)-1)